### PR TITLE
Allow per-repo and settings-based branch prefixes

### PR DESCRIPTION
**Pull request by wlonk**
tanuki

### CODE REVIEW

**Comment by wlonk** [comment]
[I want new branches to use a prefix different than the cumulusci.yml setting.](https://trello.com/c/848WB7J4/190-i-want-new-branches-to-use-a-prefix-different-than-the-cumulusciyml-setting)

File: metecho/api/gh.py [hunk 1]
```diff
@@ -147,6 +147,11 @@ def get_cumulus_prefix(**kwargs):
     """
     Expects to be in a local_github_checkout.
     """
+    branch_prefix = kwargs.pop("branch_prefix", None)
+    if branch_prefix:
+        return branch_prefix
+    if settings.BRANCH_PREFIX:
+        return settings.BRANCH_PREFIX
```
Comment (jgerigmeyer) [inline metecho/api/gh.py:154]: @wlonk I understand the theoretical value of doing this check here inside `get_cumulus_prefix`, but I might suggest moving those out into the `project_create_branch` job where this is called. A few thoughts:

1. This is no longer actually returning the `cumulus_prefix` specifically.
2. Both of these new branches (either `repository.branch_prefix` or `settings.BRANCH_PREFIX`) would not have to be called inside a `local_github_checkout`, so we could avoid that entirely in those cases.

File: metecho/api/models.py [hunk 2]
```diff
@@ -265,6 +265,7 @@ class Repository(
         validators=[validate_unicode_branch],
         default="master",
     )
+    branch_prefix = StringField(blank=True)
```
Comment (jgerigmeyer) [inline metecho/api/models.py:268]: I see many places we're using `StringField` with `null=True, blank=True`, but other places just `blank=True` -- I don't fully appreciate the distinction.

File: metecho/api/jobs.py [hunk 3]
```diff
@@ -61,6 +61,7 @@ def project_create_branch(
                 repo_url=repository.html_url,
                 repo_owner=repository.owner.login,
                 repo_branch=repository.default_branch,
+                branch_prefix=repository.branch_prefix,
```
Comment (jgerigmeyer) [inline metecho/api/jobs.py:64]: Oops; this is a naming confusion about what `repository` is in this context. It's the gh3 `repository` object, not our Metecho model.

File: metecho/api/gh.py (hunk 1 above)
Comment (wlonk) [reply metecho/api/gh.py:154]: I was having some feelings along these lines. You've convinced me.

File: metecho/api/models.py (hunk 2 above)
Comment (wlonk) [reply metecho/api/models.py:268]: Any character-type field (`StringField`, `CharField`, `TextField`, etc) can hold `None` if it is `null=True`, thus making its type `Optional[str]`, or can hold the empty string if it is `blank=True`, thus (assuming it is not also `null=True`) keeping its type `str`.

Comment (jgerigmeyer) [reply metecho/api/models.py:268]: Ok, I think I follow that. Is there a reason we're using it both ways? Is there one we prefer?

Comment (davisagli) [reply metecho/api/models.py:268]: The way I think of it is that `null` controls whether the field can store None or not, and `blank` controls whether the Django admin UI (and I guess DRF also?) will allow entering an empty string.

Comment (jgerigmeyer) [reply metecho/api/models.py:268]: I was under the impression that `null=True` was generally discouraged on text fields?

> Avoid using null on string-based fields such as CharField and TextField.

https://docs.djangoproject.com/en/3.0/ref/models/fields/#django.db.models.Field.null

**Pull request merged**